python/GafferUI/UIEditor.py

```diff
@@ -242,6 +243,35 @@ def __setColor( cls, menu, node ) :
 			with Gaffer.UndoScope( node.ancestor( Gaffer.ScriptNode ) ) :
 				Gaffer.Metadata.registerValue( node, "nodeGadget:color", color )
 
+	## Registers a custom PlugValueWidget type for a plug type, so it can be selected in the UIEditor
+	# label - String with the label to be displayed for this widget.
+	# plugClassOrTypeId - Gaffer.Plug class (or typeId) that can use the widget being registered.
+	# metadata - String with associated metadata for the widget. Usually the python identifier for the widget (ex: "GafferUI.BoolPlugValueWidget").
+	@classmethod
+	def registerPlugValueWidget( cls, label, plugClassOrTypeId, metadata ) :
+
+		# simply calls protected _PlugEditor implementation
+		_PlugEditor.registerPlugValueWidget( label, plugClassOrTypeId, metadata )
+
+	## Register additional metadata for custom PlugValueWidgets
+	# key: string with the metadata key to be set
+	# label: string with the label to be displayed for this key
+	# defaultValue: default value for the metadata if there is no current value. Also used to automatically determine the widget to use.
+	#               currently automatically supported types are strings, bools and imath.Color4f
+	# plugValueWidgetType: string with the "plugValueWidget:type" metadata value where this settings applies, or a pattern for it.
+	# presetLabelsAndValues: dictionary with { <labels>: <values> } when the widget to be used is a drop-down menu
+	# metadataWidgetCreator: callable that receives "key" and "defaultValue" arguments and returns a widget to edit the metadata
+	@classmethod
+	def registerWidgetSetting( cls, key, label, defaultValue, plugValueWidgetType, presetLabelsAndValues = None, metadataWidgetCreator = None ) :
```

Comment:
This is awkward : by attempting to hide the metadata widgets, we've ended up baking the arguments for one of them (`presetLabelsAndValues`) into a function signature which is meant to be generic. And there's no way for a caller (outside of this file) to use it, because they don't have access to the widget type in question!

What do you think to this simpler interface?

```
def registerWidgetSetting( cls, label, plugValueWidgetType, widgetCreator ) :
    ...

UIEditor.registerWidgetSetting( "Divider", "*", lambda plug : _BoolMetadataWidget( plug, "divider", False ) )
```

It's simpler, but still lets us provide any arguments we want to the custom widgets. By omitting the `key` argument, it also makes it possible for a single widget to manage multiple items of metadata, or anything else for that matter. I think maybe the only requirement is that the returned widget has a `setTarget()` method that is used when we switch the selected plug?

We could then have a separate `registerWidgetMetadata( cls, label, plugValueWidgetType, key, defaultValue )` method, to avoid overloading the one function with two different concepts. And that could be implemented simply to forward a call to `registerWidgetSetting()`.

Reply:
Unless I'm missing something, registering a `widgetCreator`, like you are suggesting, is pretty much what I was suggesting initially. The only difference I can see is that in my original proposal I would also be providing the `key`.

But I had to back away from that option because you didn't want to make the `_MetadataWidget`, and its subclasses, public.

Remember that the point of this PR is to allow us to add custom widgets and settings from outside (jabuka in this case).

All the other comments should be easy to address.

Reply:
I removed the `key` argument so that the widgets being registered didn't have to be `MetadataWidgets` or be limited to editing only one metadata. Which means we don't need to expose `MetadataWidget`, and you can implement yours how you like in Jabuka. You wanted to use the public API to make the internal registrations too, and I tried to provide a suggestion that allows you to do that, without exposing the internal widgets. It's not perfect, because of the `setTarget()` requirement, but that could be removed with some additional work (perhaps not worth the effort right now).

For the record, when I said this :

```
but backed out because I was unsure that MetadataWidget was well-thought-out enough to be promoted verbatim to the public API. An intermediate measure that I considered was to register widget metadata with its type (StringData etc) rather than with a specific widget. Would that work for you, or do you need more than the default metadata widgets?
```

My intention wasn't for you to come back with a rewrite : it was to start a discussion and decide a course of action before committing you to more work.

Reply:
> I removed the key argument so that the widgets being registered didn't have to be MetadataWidgets or be limited to editing only one metadata.

Yes. That was my understanding. I was just pointing out that was a difference to my original PR. I had also considered some way of allowing for a multi-purpose widget registration, but couldn't come up with a way that didn't require registering which widget to use directly.

> You wanted to use the public API to make the internal registrations too, and I tried to provide a suggestion that allows you to do that, without exposing the internal widgets.

Yes. Separating it into two registrations is a way to try and deal with that, and may be worth pursuing.

I just don't see how we we'd deal with scenarios where I want to use the `_MenuMetadataWidget` if we don't have the `presetLabelsAndValues` to both indicate we want that widget (without specifically referring to it) and provide the preset values to be displayed.

When first reading your comment about the two registrations, I thought you were suggesting that for those cases I'd use the direct call to `registerWidgetSetting()`, but that would require me to access `_MenuMetadataWidget` from outside.

One possibility would be for me to  re-implement that widget in Jabuka, but that seems not ideal, both because of the code duplication, as well as my impression that having widget settings that would benefit from that particular widget would be quite common, so other applications would likely to need it as well.

The other possibility is to just accept that we need the `presetLabelsAndValues`  in the `registerWidgetMetadata`. That does sound like a reasonable option to me, if we want to keep the MetadataWidgets internal.

To me, the cleanest option would be to make the MetadataWidgets public. I'm not sure what you have in mind about what their proper implementation would look like, so I don't know how much time we are talking about. But if it's significant, then we are likely to consider that not a priority at the moment. And in that case, the two registration mechanism could be a way to get something working now, but we need to figure out how to handle the `_MenuMetadataWidget` case.

> My intention wasn't for you to come back with a rewrite : it was to start a discussion and decide a course of action before committing you to more work.

Yes. It's just that, in that particular case, I decided it was easier for me to create the PR for you to see how I thought we could try to address your concerns :)

Reply:
>> You wanted to use the public API to make the internal registrations too, and I tried to provide a suggestion that allows you to do that, without exposing the internal widgets.
> Yes. Separating it into two registrations is a way to try and deal with that, and may be worth pursuing.

I had another motive for having two methods, and that was clarity. Overloading one function to do two kinds of registration makes for a more confusing API. And the `presetLabelsAndValues` made it even more so, although I understand why you added it.

> To me, the cleanest option would be to make the MetadataWidgets public.

Yeah, I'm coming to that conclusion too. I've been racking my brains to remember the source of my uneasiness about it, and come up with nothing specific. So I'll have to just put it down to a reluctance to have yet more public API to support. I'm not sure if I recall exactly how you had arranged things in your first PR. Was it a single `GafferUI/MetadataWidget.py` file with all the classes in, and them all imported into the main `GafferUI` namespace by `GafferUI/__init__.py`? Generally we have one class per file, but given how small and related these guys are, I think that's a good choice.

I think the two registrations methods I proposed (one general one for widgets, one simpler one for metadata) would still be reasonable even with publicly available MetadataWidget.

>> My intention wasn't for you to come back with a rewrite : it was to start a discussion and decide a course of action before committing you to more work.
> Yes. It's just that, in that particular case, I decided it was easier for me to create the PR for you to see how I thought we could try to address your concerns :)

I just feel bad when that work is wasted, which is why I like to ask questions first and shoot later. So, sorry, let's go back to the first approach with public widgets...

Reply:
> Was it a single GafferUI/MetadataWidget.py file with all the classes in, and them all imported into the main GafferUI namespace by GafferUI/__init__.py? Generally we have one class per file, but given how small and related these guys are, I think that's a good choice.

Yes. That was my initial suggestion. I wasn't sure about the best approach, but if you are fine with that, I'll go for it.